clang/lib/CIR/CodeGen/CIRGenExpr.cpp

```diff
@@ -2180,6 +2378,13 @@ mlir::Value CIRGenFunction::buildAlloca(StringRef name, QualType ty,
 
 mlir::Value CIRGenFunction::buildLoadOfScalar(LValue lvalue,
                                               SourceLocation Loc) {
+  return buildLoadOfScalar(lvalue.getAddress(), lvalue.isVolatile(),
+                           lvalue.getType(), getLoc(Loc), lvalue.getBaseInfo(),
+                           lvalue.isNontemporal());
+}
+
+mlir::Value CIRGenFunction::buildLoadOfScalar(LValue lvalue,
+                                              mlir::Location Loc) {
```

Comment:
Is this wrapper function required?

We should be able to get the `mlir::Location` from the given `SourceLocation`.

Reply:
@sitio-couto  I double check it, but as far as I remember I needed `mlir::Location` somewhere and the functions  `buildLoadOfScalar`  accept only `SourceLocation`.  And I just persevered the existed interface for the latter and added a couple of new functions for the former

Reply:
@gitoleg thanks for the feedback.

I took another look at this, and I think there might be an error in the call site.

In the original codegen, the call site for this function actually emits an LLVM IR load instruction:

https://github.com/llvm/clangir/blob/bb87caa687b6a035dd53463fd54131fe341d7bb1/clang/lib/CodeGen/CGExpr.cpp#L2253-L2256

To maintain parity with the original codegen, we should be building a `cir.load` in the said call site, but it is currently calling the wrapper function in question:

https://github.com/gitoleg/clangir/blob/d9df914e32d72a453fe43ce77498cc0229f99d7a/clang/lib/CIR/CodeGen/CIRGenExpr.cpp#L663-L666

The proper call to make here should be something like:

```c++
mlir::Value Val = builder.createLoad(Ptr.getPointer().getLoc(), Ptr);
```

It there a particular reason to use `buildLoadOfScalar` instead of building the operation directly with `builder.createLoad`?

Reply:
I think it's a bug indeed(